src/keyboard/KeyboardViewport.tsx

```diff
@@ -0,0 +1,153 @@
+import { FC, PropsWithChildren, useEffect, useRef } from "react";
+
+type KeyboardViewportType = PropsWithChildren<{
+  className?: string;
+}>;
+
+const KEYMAP_SCALE = "keymap:scale";
+const DEFAULT_SCALE = window.localStorage.getItem(KEYMAP_SCALE) ?? "1";
+
+export const KeyboardViewport: FC<KeyboardViewportType> = ({
+  children,
+  className,
+}) => {
+  const targetRef = useRef<HTMLDivElement>(null);
+  const scaleRef = useRef<HTMLInputElement>(null);
+
+  const setScale = (param: "increase" | "decrease") => {
+    if (!targetRef.current || !scaleRef.current) return;
+
+    const current = scaleRef.current.value;
+
+    if (param === "increase" && Number(current) < 2) {
+      scaleRef.current.value = String(Number(scaleRef.current.value) + 0.2);
+    }
+
+    if (param === "decrease" && Number(current) > 0.2) {
+      scaleRef.current.value = String(Number(scaleRef.current.value) - 0.2);
+    }
+
+    localStorage.setItem(KEYMAP_SCALE, scaleRef.current.value);
+    targetRef.current.style.setProperty(
+      "transform",
+      `scale(${scaleRef.current.value})`,
+    );
+  };
+
+  const resetScale = () => {
+    if (!targetRef.current || !scaleRef.current) return;
+    targetRef.current.style.translate = "unset";
+    targetRef.current.style.setProperty("transform", "scale(1)");
+    scaleRef.current.value = "1";
+    localStorage.setItem(KEYMAP_SCALE, "1");
+  };
+
+  useEffect(() => {
+    if (!targetRef.current) return;
+
+    const target = targetRef.current;
+    const offset = { x: 0, y: 0 };
+    let isPanningActive = false;
+
+    function panStart(e: KeyboardEvent) {
```

Comment:
So this only supports pan via spacebar? I think that's a good addition, but I think this should support click and drag, as most users will use that.

Reply:
we can add that as well.